### ISSUE TITLE
Converte ElementFragmentAlgorithm to use smart pointers

### ISSUE THREAD

**Issue by bwspenc** [idaholab/moose#6417]
### Description of the enhancement or error report

The ElementFragmentAlgorithm used by XFEM does quite a bit of dynamic allocation using new/delete. Refactor this to use smart pointers.

### Rationale for the enhancement or information for reproducing the error

Avoid potential memory leaks

### Identified impact

Extensive changes to ElementFragmentAlgorithm and associated classes

**Comment by permcody** [comment]
Looks like this is pretty important. Almost all of your XFEM tests are leaking memory.

**Comment by dschwen** [comment]
![b1f](https://cloud.githubusercontent.com/assets/202302/13356718/16c0c5d6-dc64-11e5-8004-b54cad5dd890.gif)

**Comment by bwspenc** [comment]
Inconceivable!

**Comment by bwspenc** [comment]
It actually looks like the memory leaks are caused by the use of MooseArray objects. They don't appear to delete their memory.

**Comment by friedmud** [comment]
@bwspenc They don't delete their memory _automatically_ on purpose... because the memory is shared between multiple MooseArrays... so deleting it in one would leave dangling pointers in the others.

You can tell it to delete the memory explicitly if you know it's safe by using http://mooseframework.org/docs/doxygen/moose/classMooseArray.html#afb89810e7050acc276831d7a802b4e66

MooseArrays are special beings... they are there specifically to enable the kind of memory sharing/moving we need to do inside the library.  That said: they will certainly get MUCH better (possibly even eliminated!) when we move to C++11...

**Comment by bwspenc** [comment]
@friedmud Yeah, I'm sitting next to David, and he explained that to me.  I just blindly used MooseArrays thinking they would behave like std::vectors.  I added a release(), so I think all is good now as far as that goes.
